transformer_engine/common/transformer_engine.cpp

```diff
@@ -273,6 +273,284 @@ void CheckOutputTensor(const Tensor &t, const std::string &name, bool allow_empt
+    NVTE_CHECK(t.columnwise_scale_inv.num_tensors == expected_num_tensors, "Grouped tensor ", name,
+               " columnwise_scale_inv.num_tensors (", t.columnwise_scale_inv.num_tensors,
+               ") doesn't match expected num_tensors (", expected_num_tensors, ")");
+  }
+  if (t.amax.has_data()) {
+    NVTE_CHECK(t.amax.num_tensors == expected_num_tensors, "Grouped tensor ", name,
+               " amax.num_tensors (", t.amax.num_tensors, ") doesn't match expected num_tensors (",
+               expected_num_tensors, ")");
+  }
+  if (t.columnwise_amax.has_data()) {
+    NVTE_CHECK(t.columnwise_amax.num_tensors == expected_num_tensors, "Grouped tensor ", name,
+               " columnwise_amax.num_tensors (", t.columnwise_amax.num_tensors,
+               ") doesn't match expected num_tensors (", expected_num_tensors, ")");
+  }
+  if (t.scale.has_data()) {
+    NVTE_CHECK(t.scale.num_tensors == expected_num_tensors, "Grouped tensor ", name,
+               " scale.num_tensors (", t.scale.num_tensors,
+               ") doesn't match expected num_tensors (", expected_num_tensors, ")");
+  }
+
+  // Check that shape information is provided and consistent
+  NVTE_CHECK(t.first_dims.dptr != nullptr, "Grouped tensor ", name,
+             " must have first_dims allocated");
+  NVTE_CHECK(t.second_dims.dptr != nullptr, "Grouped tensor ", name,
+             " must have second_dims allocated");
+
+  // Ensure shape arrays are 1D
+  NVTE_CHECK(t.first_dims.shape.size() == 1, "Grouped tensor ", name,
+             " first_dims must be 1D (got ", t.first_dims.shape.size(), "D)");
+  NVTE_CHECK(t.second_dims.shape.size() == 1, "Grouped tensor ", name,
+             " second_dims must be 1D (got ", t.second_dims.shape.size(), "D)");
+
+  // Ensure shape arrays have correct dtype (size_t = Int64)
+  NVTE_CHECK(t.first_dims.dtype == DType::kInt64, "Grouped tensor ", name,
+             " first_dims must have dtype Int64 (got ", to_string(t.first_dims.dtype), ")");
+  NVTE_CHECK(t.second_dims.dtype == DType::kInt64, "Grouped tensor ", name,
+             " second_dims must have dtype Int64 (got ", to_string(t.second_dims.dtype), ")");
+
+  NVTE_CHECK(!t.first_dims.shape.empty() && t.first_dims.shape[0] > 0, "Grouped tensor ", name,
+             " first_dims has invalid shape");
+  NVTE_CHECK(!t.second_dims.shape.empty() && t.second_dims.shape[0] > 0, "Grouped tensor ", name,
+             " second_dims has invalid shape");
+  NVTE_CHECK(t.first_dims.shape[0] == t.second_dims.shape[0], "Grouped tensor ", name,
+             " first_dims and second_dims must have the same size (got ", t.first_dims.shape[0],
+             " vs ", t.second_dims.shape[0], ")");
+  NVTE_CHECK(t.first_dims.shape[0] == t.num_tensors(), "Grouped tensor ", name,
+             " shape arrays size (", t.first_dims.shape[0], ") doesn't match num_tensors (",
+             t.num_tensors(), ")");
+
+  // Check cumulative_tensor_sizes if it's set (optional but must be consistent)
```

Comment:
Why would it be optional? We need to know the offsets to be able to do anything interesting and for that we need the allocation.